### PR TITLE
CAN Assert Testing

### PR DESCRIPTION
**Pull request by jackberezny**
- Delays were for alumni demo to ensure that all statements print
- select_mob() still has to be debugged (errors integrating with assert_can)
- TO DO: need to add id_tag, id_mask, callback functions

### CODE REVIEW

**Comment by brunofalmeida** [comment]
Should this pull request be named something related to testing CAN?

**Comment by siddMahen** [comment]
@jackberezny Recently finished the first version of the test harness, see #28. Could you fork that branch and rewrite these tests using the new framework? Check out tests/arithmetic and tests/print for examples.

**Comment by jackberezny** [comment]
@siddMahen Let me know what you think, I think it's getting close to being ready to be merged. 
Just a couple questions:
- Should the file be named a variation of "main.c"?
- Should ASSERT_TRUE be changed to ASSERT_EQ when either works?

**Comment by siddMahen** [comment]
@jackberezny This exactly the format I'm after, great job. The file should be named `main1.c`. If you're comparing two 8-bit integer quantities, then ASSERT_EQ is preferred; you can often reduce to this case by bit-shifting and masking. Be sure to test this on a board, and let me know when it is final.

**Comment by siddMahen** [comment]
In general, you want to avoid things like `ASSERT_TRUE(A == B)`; in this situation it's much better to use `ASSERT_EQ(A, B)`.

**Comment by siddMahen** [comment]
@jackberezny Has this been tested?

File: tests/assert_can/main1.c [hunk 1]
```diff
@@ -47,7 +47,7 @@ void rx_callback(const uint8_t* data, uint8_t len) {
 void tx_callback(uint8_t* data, uint8_t* len) {
     *len = 4;
     print("data: ");
-    for (int i = 0; i < len; i++){
+    for (int i = 0; i < *len; i++){
```
Comment (siddMahen) [inline tests/assert_can/main1.c]: This is very weird; better to just write 4, if you know that's the length

**Comment by siddMahen** [comment]
@jackberezny Has this been tested?

**Comment by jackberezny** [comment]
@siddMahen I tried to test last night but had some issues with the test harness. I’m going to the lab again on Thursday morning so hopefully I can get it working then.

**Comment by siddMahen** [comment]
@jackberezny Hmm, ok. Try testing this on a Mac; for technical reasons, Windows is giving us trouble. If there are any unknown bugs in the harness, please let me know.

**Comment by jackberezny** [comment]
@siddMahen  CAN Assert has been tested and works without error; should be ready to be merged after commit history is cleaned up.

**Comment by siddMahen** [comment]
@jackberezny Great work! I agree, this should be merged after commit cleanup and small fixes. I think the `/examples/assert_can` can be safely deleted. Looking into the future, it would be nice to have some "negative" tests. For example, make sure that something that should fail/create an error definitely fails and gives the right error.

**Comment by brunofalmeida** [comment]
@jackberezny @siddMahen Is this ready to be merged?

**Comment by siddMahen** [comment]
@brunofalmeida This needs to be tested on a board, and the commits need to be rebased, otherwise it should be good to go.

**Comment by siddMahen** [comment]
@jackberezny Looking good! A few minor things: commit messages should begin with a title, and each line should contain at most 80 characters; this makes commits looks pretty in (almost all) terminals. I'm ready to merge this as soon as you let me know it's been tested on a board.

**Comment by jackberezny** [comment]
@siddMahen I tested the code before qual and it worked as expected. Is there any reason to test again?

**Comment by siddMahen** [comment]
@jackberezny Great work! Glad to see you're becoming more familiar with `git rebase`! Combing through this one last time, should be ready to merge today.

**Comment by siddMahen** [comment]
@jackberezny Removed the `print()` statements, which were unnecessary, rebased on top of master and made some minor style changes.

**Comment by siddMahen** [comment]
@jackberezny Just as a precaution, I tested this on the old boards and ran into an error:

```
./bin/harness.py -p /dev/tty.usbmodem00187462 /dev/tty.usbmodem00208212 -d tests
WARNING: The UART TX pin(s) on SCK must be disconnected before upload.
Ensure the UART TX pin on each board is disconnected before proceeding.
Run test suite 'can'? (y/n) y
    Compiling...
    Linking...
    Uploading...
Connect the UART TX pin(s) to SCK. (ok) ok
--------------------------------------------------------------------------------
Test: init_can
Test complete - Passed
--------------------------------------------------------------------------------
Test: init_tx
Test complete - Passed
--------------------------------------------------------------------------------
Test: init_rx
Test complete - Passed
--------------------------------------------------------------------------------
Test: pause/resume
data_s:   0  1  2  3
TX sent
ERR: Bad DLC.
TX received!
data_s:   0  1  2  3
TX sent
In function 'pause_resume_test', line 129
    Error: ASSERT_EQ failed
Test complete - Failed
--------------------------------------------------------------------------------

Test: error handling
data_s:   0  1  2  3
TX sent
TX received!
Test complete - Passed
--------------------------------------------------------------------------------
Test suite 'can' complete
    Time elapsed: 1.797 s
    Passed: 4 / 5
    Failed: 1 / 5
--------------------------------------------------------------------------------
```

The offender is line 129 of `main1.c1`:

```
    ...
    // Verifying resumption and CANCDMOB register contents
    resume_mob(&tx_mob);
    ASSERT_EQ(CANCDMOB & 0xC0, 0x80);
    ...
```

**Comment by jackberezny** [comment]
@siddmahen thanks for testing it. I think that line might supposed to have 0x40 instead of 0x80. Hopefully that fixes the problem.

**Comment by siddMahen** [comment]
Ok, will try that and merge if possible.

**Comment by siddMahen** [comment]
@jackberezny Congratulations! Everything checks out; merging now.

**Pull request merged**